Create UserTask (discover-rds) when RDS is missing IAM Authentication

marcoandredinis opened the pull request:
This PR changes the Discovery Service to start creating UserTasks (of discover-rds type) when the discovered RDS database does not have IAM DB Authentication enabled.

Demo:
Setup: two databases are found but only one has IAM DB Authentication enabled:
`$ tctl get user_tasks`
```yaml
kind: user_task
metadata:
  expires:
    nanos: 330312000
    seconds: 1736856115
  name: d2e7cb28-03b1-58ef-9816-08e7ade0a85b
  revision: c02cadd4-1ab9-43ac-a2fe-eeca8238f893
spec:
  discover_rds:
    account_id: "278576220453"
    databases:
      marco-discovertask-02:
        discovery_config: 8cdad48f-6f49-4f37-b333-86a310122f6f
        discovery_group: aws-prod
        engine: aurora-postgresql
        is_cluster: true
        name: marco-discovertask-02
        sync_time:
          nanos: 325158000
          seconds: 1736855515
    region: eu-west-2
  integration: teleportdev
  issue_type: rds-iam-auth-disabled
  state: OPEN
  task_type: discover-rds
status:
  last_state_change:
    nanos: 132958000
    seconds: 1736855001
version: v1

marcoandredinis commented on lib/srv/discovery/database_watcher.go:
```diff
@@ -134,11 +139,51 @@ func (s *Server) startDatabaseWatchers() error {
+		return
+	}
+
+	s.Log.WarnContext(s.ctx, "IAM POLICY STATUS FOR", "db", db.GetName(), "policy_status", db.GetAWS().IAMPolicyStatus.String(), "rds", db.GetAWS().RDS)
```
😱

marcoandredinis commented on lib/srv/discovery/database_watcher.go:
```diff
@@ -134,11 +139,51 @@ func (s *Server) startDatabaseWatchers() error {
 			for dc := range discoveryConfigsChanged {
 				s.updateDiscoveryConfigStatus(dc)
 			}
+			s.upsertTasksForAWSRDSFailedEnrollments()
```
No. How often it gets called depends on the configured matcher and the User's RDS databases.
Every agent running the same discovery group and integration, should generate the same UserTasks because it will also see the same databases.

marcoandredinis commented on lib/srv/discovery/status.go:
```diff
@@ -752,3 +814,103 @@ func (s *Server) discoverEKSUserTaskAddExistingClusters(currentUserTask *usertas
+	s.awsRDSTasks.mu.Lock()
+	defer s.awsRDSTasks.mu.Unlock()
+	for g := range s.awsRDSTasks.issuesSyncQueue {
+		if err := s.mergeUpsertDiscoverRDSTask(g, s.awsRDSTasks.databaseIssues[g]); err != nil {
```
It should be ok because `issuesSyncQueue` is pretty small, so I don't expect it to take that long to merge.

It is necessary to prevent issues being lost and/or being reported while the actual issue was fixed.
I can call a goroutine here and do it async, but then we might be reporting issues that are no longer occurring or taking a little more time to report issues.

GavinFrazar reviewed (approved):
LGTM but I wonder if the 5 second lease expiry is too short in `acquireSemaphoreForUserTask` ?

marcoandredinis commented on lib/srv/discovery/status.go:
```diff
@@ -752,3 +814,103 @@ func (s *Server) discoverEKSUserTaskAddExistingClusters(currentUserTask *usertas
+func (s *Server) upsertTasksForAWSRDSFailedEnrollments() {
+	s.awsRDSTasks.mu.Lock()
+	defer s.awsRDSTasks.mu.Unlock()
+	for g := range s.awsRDSTasks.issuesSyncQueue {
```
`databaseIssues` has all the currently known issues for this iteration.
Some of those are already synced to the cluster, but others were not.

`issuesSyncQueue` only contains the issues that were not synced.
Iterating over `issuesSyncQueue` ensures we don't update UserTasks that were not changed.

marcoandredinis replied:
Let's say we are enrolling databases from two regions.
The watcher sends databases from region A (`case newResources := <-watcher.ResourcesC():`) and we detect 1 database which is missing the IAM DB Auth.
We call `addFailedEnrollment` for that database and it creates a UserTask in memory.
We then call `upsertTasksForAWSRDSFailedEnrollments` and at this point, you are correct that both `issuesSyncQueue` and `databaseIssues` have the same keys.
After merging, `issuesSyncQueue` has 0 keys but `databaseIssues` still has the key.

Now, lets say that watcher sends more databases but this time for region B.
Given it's a different region, it must be another UserTask.
So we call `addFailedEnrollment`, and another task is created.
`issuesSyncQueue` has 1 key (for region B) but `databaseIssues` has two keys.

Calling `upsertTasksForAWSRDSFailedEnrollments` will loop over `issuesSyncQueue` which means only this new task is sent to the Cluster.
However, if we used `databaseIssues`, then we would be sending the issue for region B, and for region A (which was not changed at all).

We can't also clear `databaseIssues`, after `upsertTasksForAWSRDSFailedEnrollments`, because we might receive more databases for region A.

marcoandredinis replied:
The watcher can send multiple resources using the `watcher.ResourcesC()` in a single iteration. So, while it's true that there's a reset when the iteration starts, I don't think we can assume only one item will be sent over the channel per iteration.

Also, EC2 and EKS watchers are using the same logic. So, even if this somewhat wasted computation I think it's worth it so that we keep consistency with the other watchers.

Given that you approved and that this seems like an optimization (at best) I will go ahead and merge it.

marcoandredinis commented:
> LGTM but I wonder if the 5 second lease expiry is too short in `acquireSemaphoreForUserTask` ?

Updated to 10s.



> Oh, also can we get the times in your example to format in human readable format instead of seconds/nanoseconds? (if that's a tctl output)?

That's a tctl output.
I will do it in another PR.
This should be the same for all UserTasks (not only for the database sync time but also for metadata.expires).

marcoandredinis committed:
Create UserTask (discover-rds) when RDS is missing IAM Authentication

This PR changes the Discovery Service to start creating UserTasks (of
discover-rds type) when the discovered RDS database does not have IAM DB
Authentication enabled.
